From 1d9315f1d220c1b980198a5a89f855eac692fdcd Mon Sep 17 00:00:00 2001
From: Preston Petrie <preston159@users.noreply.github.com>
Date: Tue, 13 May 2025 16:13:15 -0500
Subject: [PATCH 6/6] add rpm-ostree wrapper scripts

Added rpm-ostree-update, rpm-ostree-install, and rpm-ostree-uninstall
scripts which perform an update and, in the latter two cases, install
or uninstall layered packages. These scripts save a log file with
package changes to ~/rpm-ostree-log/. Additionally, a wrapper around
rpm-ostree itself is added to automatically run one of the other wrapper
scripts when rpm-ostree [update|install|uninstall] is run.
---
 .../overrides/usr/bin/.rpm-ostree-create-diff      | 12 ++++++++++++
 system_files/overrides/usr/bin/rpm-ostree          | 14 ++++++++++++++
 system_files/overrides/usr/bin/rpm-ostree-install  |  5 +++++
 .../overrides/usr/bin/rpm-ostree-uninstall         |  5 +++++
 system_files/overrides/usr/bin/rpm-ostree-update   |  4 ++++
 5 files changed, 40 insertions(+)
 create mode 100755 system_files/overrides/usr/bin/.rpm-ostree-create-diff
 create mode 100755 system_files/overrides/usr/bin/rpm-ostree
 create mode 100755 system_files/overrides/usr/bin/rpm-ostree-install
 create mode 100755 system_files/overrides/usr/bin/rpm-ostree-uninstall
 create mode 100755 system_files/overrides/usr/bin/rpm-ostree-update

diff --git a/system_files/overrides/usr/bin/.rpm-ostree-create-diff b/system_files/overrides/usr/bin/.rpm-ostree-create-diff
new file mode 100755
index 00000000..d1b168fd
--- /dev/null
+++ b/system_files/overrides/usr/bin/.rpm-ostree-create-diff
@@ -0,0 +1,12 @@
+#!/usr/bin/zsh
+
+DIR="$HOME/rpm-ostree-log"
+mkdir -p $DIR
+
+VERSIONS=$(/bin/rpm-ostree status | \grep -i Version: | head -2 | xargs -n3 | cut -d' ' -f2 | sort | xargs)
+FILE="$DIR/${VERSIONS/ /-}$1"
+touch $FILE
+date >> $FILE
+echo $2 >> $FILE
+/bin/rpm-ostree db diff >> $FILE
+echo "diff file: $FILE"
diff --git a/system_files/overrides/usr/bin/rpm-ostree b/system_files/overrides/usr/bin/rpm-ostree
new file mode 100755
index 00000000..ef7bec1f
--- /dev/null
+++ b/system_files/overrides/usr/bin/rpm-ostree
@@ -0,0 +1,14 @@
+#!/usr/bin/zsh
+
+if [[ "update" == "$1" ]]; then
+  shift
+  rpm-ostree-update $@
+elif [[ "install" == "$1" ]]; then
+  shift
+  rpm-ostree-install $@
+elif [[ "uninstall" == "$1" ]]; then
+  shift
+  rpm-ostree-uninstall $@
+else
+  /bin/rpm-ostree $@
+fi
diff --git a/system_files/overrides/usr/bin/rpm-ostree-install b/system_files/overrides/usr/bin/rpm-ostree-install
new file mode 100755
index 00000000..848572b7
--- /dev/null
+++ b/system_files/overrides/usr/bin/rpm-ostree-install
@@ -0,0 +1,5 @@
+#!/usr/bin/zsh
+
+PACKAGES=$(sed -E -e 's/^| /&--install=/g' <<< "$@")
+/bin/rpm-ostree update ${=PACKAGES}
+"${0:a:h}/.rpm-ostree-create-diff" "-install" "install $@"
diff --git a/system_files/overrides/usr/bin/rpm-ostree-uninstall b/system_files/overrides/usr/bin/rpm-ostree-uninstall
new file mode 100755
index 00000000..93cfdf27
--- /dev/null
+++ b/system_files/overrides/usr/bin/rpm-ostree-uninstall
@@ -0,0 +1,5 @@
+#!/usr/bin/zsh
+
+PACKAGES=$(sed -E -e 's/^| /&--uninstall=/g' <<< "$@")
+/bin/rpm-ostree update ${=PACKAGES}
+"${0:a:h}/.rpm-ostree-create-diff" "-uninstall" "uninstall $@"
diff --git a/system_files/overrides/usr/bin/rpm-ostree-update b/system_files/overrides/usr/bin/rpm-ostree-update
new file mode 100755
index 00000000..579a7679
--- /dev/null
+++ b/system_files/overrides/usr/bin/rpm-ostree-update
@@ -0,0 +1,4 @@
+#!/usr/bin/zsh
+
+/bin/rpm-ostree update $@
+"${0:a:h}/.rpm-ostree-create-diff"
-- 
2.49.0

